weasyl/test/login/test_authenticate_bcrypt.py

```diff
@@ -0,0 +1,140 @@
+    result = login.authenticate_bcrypt(username='', password='password')
+    assert result == (0, 'invalid')
+
+
+def test_logins_fail_if_no_password_provided():
+    user_name = "NoPasswordProvided"
+    result = login.authenticate_bcrypt(username=user_name, password='')
+    assert result == (0, 'invalid')
+
+
+def test_login_fails_if_incorrect_username_is_provided():
+    user_name = "InvalidUsernameProvided"
+    result = login.authenticate_bcrypt(username=user_name, password=raw_password)
+    assert result == (0, 'invalid')
+
+
+def test_login_fails_for_incorrect_credentials():
+    user_name = "testAuthBcry0001"
+    random_password = "ThisIsARandomPassword"
+    db_utils.create_user(password=random_password, username=user_name)
+    another_random_password = "ThisIsNotTheSamePassword"
+    result = login.authenticate_bcrypt(username=user_name, password=another_random_password)
+    assert result == (0, 'invalid')
+
+
+def test_login_fails_for_invalid_auth_and_logs_failure_if_mod_account(tmpdir, monkeypatch):
+    # Set the temporary directory for the log file
+    monkeypatch.setenv(macro.MACRO_SYS_LOG_PATH, tmpdir + "/")
+    log_path = '%s%s.%s.log' % (macro.MACRO_SYS_LOG_PATH, 'login.fail', d.get_timestamp())
+    user_id = db_utils.create_user(username='ikani', password=raw_password)
+    # Set the moderators in libweasyl
+    configure_moderators(mod_user_id=user_id)
+    # Ensure we are actually writing to the file by counting the file's lines
+    prerun_loglines = 0
+    # The file might not exist; this is fine; ignore
+    try:
+        with open(log_path, 'r') as log:
+            for line in log:
+                prerun_loglines += 1
+            log.close()
+    except IOError:
+        pass
+    postrun_loglines = 0
+    # Item under test
+    result = login.authenticate_bcrypt(username='ikani', password='FakePassword')
+    # Verify we are writing to the log file as expected
+    last_line = ''
+    with open(log_path, 'r') as log:
+        for line in log:
+            postrun_loglines += 1
+        last_line = line
+        log.close()
+    last_line_dict = json.loads(last_line)
+    assert postrun_loglines > prerun_loglines
+    assert last_line_dict['userid'] == user_id
+    assert result == (0, 'invalid')
+
+
+def test_login_fails_if_user_is_banned():
+    user_name = "testAuthBcry0002"
+    user_id = db_utils.create_user(password=raw_password, username=user_name)
+    d.engine.execute("UPDATE login SET settings = 'b' WHERE userid = %(id)s", id=user_id)
+    result = login.authenticate_bcrypt(username=user_name, password=raw_password)
+    assert result == (user_id, 'banned')
+
+
+def test_login_fails_if_user_is_suspended():
+    user_name = "testAuthBcry0003"
+    user_id = db_utils.create_user(password=raw_password, username=user_name)
+    d.engine.execute("UPDATE login SET settings = 's' WHERE userid = %(id)s", id=user_id)
+    release_date = d.convert_unixdate(31, 12, 2030)
```

Comment:
I don’t think we’re mocking the date, so to avoid having this break in 2030 when weasyl3 is almost done, use `d.get_time() + 60`?